backend/controller/controller.go

```diff
@@ -929,3 +966,19 @@ func extractIngressRoutingEntries(req *ftlv1.CreateDeploymentRequest) []dal.Ingr
 	}
 	return ingressRoutes
 }
+
+func (s *Service) getLogger(ctx context.Context, deploymentKey *model.DeploymentKey) *log.Logger {
+	if deploymentKey == nil {
+		return s.logger
+	}
+
+	deploymentLogger, ok := s.deploymentLoggers[*deploymentKey]
+	if ok {
+		return deploymentLogger
+	}
+
+	// create deployment logger if it doesn't exist
+	logger := log.FromContext(ctx).AddSink(newDeploymentLogsSink(s.ctx, *deploymentKey, s.dal))
+	s.deploymentLoggers[*deploymentKey] = logger
```

Comment:
This is not thread safe. You could either use a `sync.Map`, or wrap all access (reads and writes) in a `sync.Mutex`.

Reply:
You'll also need some system for deleting them, which I don't see here?